### DISCUSSION TITLE
Translate Organic Maps website on Weblate

### DISCUSSION THREAD

**Discussion by rtsisyk** [organicmaps/organicmaps#4076] [Translations]
Guten Tag @organicmaps/translations teams, 

It has been a while since our previous i18n initiative (https://github.com/organicmaps/organicmaps/discussions/1489). Thanks to everyone who contributed and helped us to make Organic Maps available in more languages. Now we have translation teams for popular languages, there are still a lot of vacant places. Please use this opportunity to get the maintainer hat of your native languages on Organic Maps. Here is a link: https://github.com/organicmaps/organicmaps/discussions/1489.

Today I am glad to announce the new [**experimental**](https://github.com/organicmaps/organicmaps.github.io/blob/master/TRANSLATIONS.md#how-it-works) integration of [organicmaps.app](https://organicmaps.app) website with [Weblate](https://hosted.weblate.org/projects/organicmaps/website/). The new [**experimental**](https://github.com/organicmaps/organicmaps.github.io/blob/master/TRANSLATIONS.md#how-it-works) system can process [Markdown content](https://github.com/organicmaps/organicmaps.github.io/tree/i18n/content) from static site generator (SSG) into individual `gettext` localizable strings handled by Weblate. Now it is possible to keep all languages in sync and use all rich features of Weblate to translate the website. Every new sentence or paragraph added to English version will be automagically propagated to other languages. It sounds like a magic, and it is. Thanks F-Droid team for this brilliant idea and implementation.

As usual, any approach has its pros and cons. The primary downside of the new system is that the structure of content will be more rigid and should follow English version by the structure. It is still possible to reword sentences in any preferable way, add multiple sentences instead of one, have custom links, etc. However, changing the order of paragraphs is probably not a good idea. But you can use your favorite Weblate tool, get notified for the new content, use machine translation tooling and last but not least—focus on **translations**, not formatting and syntax. 

The only right solution to solve this dilemma is to ask you, the people who work with translations every day, for your advice. 

Let's do the pilot project and evaluate this new approach to see how it flies.

During evaluation period the following existing languages below will **continue to use the old raw Markdowns**:

- German
- French
- Italian
- Polish
- Turkish
- Russian

Please continue to edit Markdowns directly in GitHub (and have this hassle of maintaining content in-sync with English) for languages above. It is completely OK to play with Weblate to get some hands-on experience with the new system, but we don't have a good way to support edits from both sides (GitHub and Weblate) at the same time.

For all other new languages, please feel free to press a button on [Weblate](https://hosted.weblate.org/projects/organicmaps/website/) and start translation. As soon as 80% translation threshold is reached—the language will be added to [organicmaps.app](https://organicmaps.app) website. Even if we call this experiment failed later, all new translations from Weblate will go to GitHub and preserved. You can use any other tools that can handle `.po` gettext files, the integration is not limited to Weblate.

Useful links:

- [Weblate Project](https://hosted.weblate.org/projects/organicmaps/website/)
- [Detailed instructions](https://github.com/organicmaps/organicmaps.github.io/blob/i18n/TRANSLATIONS.md#how-it-works)

 Any feedback and opinions are welcomed! The decision will be made purely based on this pilot evaluation and opinion of the most active contributors.

**Comment by rtsisyk** [comment]
Thank @Linerly for contributing Indonesian: https://hosted.weblate.org/changes/?project=organicmaps&component=website&lang=id

**Reply by biodranik** [reply]
Looks like localized references/urls should be added to properly handle `[berdonasi][donasite]` instead of `[donate][donate]`. Even better would be to embed URLs into the original text, then they can be easily localized like `[berdonasi](/id/donate)`

**Reply by Linerly** [reply]
I did try to rename the translated references/URLs to the original one (and fix some Markdown as well) due to machine translation errors

**Reply by rtsisyk** [reply]
> I did try to rename the translated references/URLs to the original one (and fix some Markdown as well) due to machine translation errors

I think that it should work. I will try to compile it tomorrow and push a branch to GitHub with generated Markdowns.

**Reply by rtsisyk** [reply]
Indonesian has been integrated and now available on the website. This is the first language we translated via Weblate.
@Linerly, thanks a lot for contribution. Please tell us about your experience of using Weblate. Was it convenient or not? What difficulties did you have?

**Reply by Linerly** [reply]
Using Weblate is very much convenient as I am used to using it for translating projects, rather than having to fork, clone the repository locally, translate by having to edit the files carefully, commit and push, then make a pull request — I can just translate just using a browser

I can even do it on the go, though the interface isn't the best on mobile sometimes — still better than having to do it manually

Weblate also sends an email whenever a string is added or updated, so I can translate it immediately

I also have used several localization platforms other than Weblate, but I think Weblate is the best when it comes to search and replacing words inside of strings, and has many advanced features too (including translating offline by downloading either the whole or just the unfinished strings)

**Reply by rtsisyk** [reply]
Published.

**Comment by jimcarst** [comment]
Tried adding Dutch (Nederlands), but with no success. If someone else can add this language, I'll translate.

**Reply by rtsisyk** [reply]
Hi, Dutch is already added: https://hosted.weblate.org/projects/organicmaps/website/nl/. Please try.

**Reply by jimcarst** [reply]
Thanks. Done.

**Reply by rtsisyk** [reply]
Published

**Comment by delthia** [comment]
Hi,
I tried adding spanish, but I can't figure out how to get that working. I downloaded de files and created a new `.po` file with the spanish translations and it is now almost fully translated.
How should I proceeed? Should I open a pull request or could the file be uploaded to weblate once the language is oppened?

Thanks

**Reply by rtsisyk** [reply]
Hi, just use Weblate https://hosted.weblate.org/projects/organicmaps/website/es/ for translations.

**Reply by delthia** [reply]
Thanks.
I'll add the translations there.
Before I couldn't, as the language wasn't added and the add button didn't make a request, like it does for some other projects in Weblate.

**Reply by rtsisyk** [reply]
Published

**Comment by matheusgomesms** [comment]
First of all: congrats of moving this part to be translated on Weblate!

Just to give a feedback: I did the translation of Brazilian Portuguese. It was so easy! Of course, maybe more instructions are needed for some URLs, but overall it was a great experience (not my first time using Weblate or Transifex, though).

I understand the challenges, but I REALLY hope that whole OMaps app can be translated on these platforms.

**Reply by rtsisyk** [reply]
Thanks for contributing Brazilian Portuguese!!! I added 17 more strings that are not handled automatically, like menu items, image captions and others. Please take a look. We don't have the full integration for those 17 strings yet because of totally missing i18n in Zola, but this is just matter of time.

**Reply by rtsisyk** [reply]
@matheusgomesms is this new Weblate integration more convenient than translation and synchronization of individual Markdown files on GitHub? We need your feedback. It is unrealistic to support both methods with bidirectional synchronization.

**Reply by matheusgomesms** [reply]
Waaay better! Really, I always have second thoughts when translation here on Github, but on Weblate is a no brainer.

(I will add the new translations soon, thanks!)

**Reply by rtsisyk** [reply]
Published

**Comment by rtsisyk** [comment]
2022-12-19 status update:

|  Language |  %  | Published | Kudos to | 
|:--|:-:|:-:|---|
|   [Indonesian](https://hosted.weblate.org/projects/organicmaps/website/id/)  | 99%  | ✅ |  @Linerly    |  
|  [Portuguese (Brazil)](https://hosted.weblate.org/projects/organicmaps/website/pt_BR/)  |  **89%**  | ⏳  | @matheusgomesms |   
|  [Basque](https://hosted.weblate.org/projects/organicmaps/website/eu/)  |  **89%**  | ⏳  |  @garaolaza  |  
|  [Dutch](https://hosted.weblate.org/projects/organicmaps/website/nl/)   |  26% |   |  owje at proton <br/> @jimcarst is it you?  |  
|  [Ukrainian](https://hosted.weblate.org/projects/organicmaps/website/uk/)   |  7% |   |   @Kefir2105   |  


Note: Basque and Portuguese were 100%, but we added 17 new strings which are still not integrated with Markdown/Zola automatically.

**Reply by rtsisyk** [reply]
2022-12-23 status update:

- Integrated Zola's `config.toml` and menu items into Weblate
- Now 100% of website content can be translated via Weblate
- New languages can be added directly via Weblate
- Fixed broken links on several pages
- Published Portuguese (Brazil), Dutch, Spanish, Basque

|  Language |  %  | Published | Kudos to | 
|:--|:-:|:-:|---|
|   [Indonesian](https://hosted.weblate.org/projects/organicmaps/website/id/)  | 100%  | [ ✅](https://organicmaps.app/id/) |  @Linerly    |  
|  [Portuguese (Brazil)](https://hosted.weblate.org/projects/organicmaps/website/pt_BR/)  |  100%  |  [ ✅](https://organicmaps.app/pt-BR/)  | @matheusgomesms |   
|  [Dutch](https://hosted.weblate.org/projects/organicmaps/website/nl/)   |  99% | [ ✅](https://organicmaps.app/nl/)   |  [@appletree](https://hosted.weblate.org/user/appletree/)  |  
|  [Spanish](https://hosted.weblate.org/projects/organicmaps/website/es/)  |  98%  | [ ✅](https://organicmaps.app/es/)   | @delthia |   
|  [Basque](https://hosted.weblate.org/projects/organicmaps/website/eu/)  |  96%  | [ ✅](https://organicmaps.app/eu/)   |  @garaolaza  |  
|  [Ukrainian](https://hosted.weblate.org/projects/organicmaps/website/uk/)   |  27% |   |   @Kefir2105   |

**Comment by rtsisyk** [comment]
Kudos to @Grreby for [contributing Czech](https://github.com/organicmaps/organicmaps.github.io/pull/70) (without Weblate)! 

We are evaluating this Weblate integration as a replacement to the current approach of manually translating and synchronizing raw Markdowns on GitHub. It is unrealistic to support both methods with bidirectional synchronization. @Grreby, it would be nice if you can take a look at [Weblate](https://hosted.weblate.org/projects/organicmaps/website/cs/) and give us your feedback if you have time. No need to duplicate all translations, just play with a couple of strings get some experience of this new alternative tooling.

**Reply by rtsisyk** [reply]
@Grreby did you have a chance to take a look at [Weblate](https://hosted.weblate.org/projects/organicmaps/website/cs/)?

**Reply by Grreby** [reply]
Hi, I am so sorry I missed this conversation. 

I think that the current way of translating is kinda bad, because many people who are willing to translate aren't developers. That means that they may have problems with Git, pull requests etc. Because of this, I think Weblate is a much better choice. However, when I look at the Czech translations there, it seems like its auto-translated and not actually the original translations. Is that possible? 

Thanks for your hard work!

**Reply by rtsisyk** [reply]
> Hi, I am so sorry I missed this conversation.
> 
> I think that the current way of translating is kinda bad, because many people who are willing to translate aren't developers. That means that they may have problems with Git, pull requests etc. Because of this, I think Weblate is a much better choice. However, when I look at the Czech translations there, it seems like its auto-translated and not actually the original translations. Is that possible?
> 
> Thanks for your hard work!

Unfortunately, there are no any automated way of extraction translations from Markdowns into Weblate. It works only in the opposite direction - translated strings from Weblate can be used to re-generate Markdown. I am afraid that the only way of moving from Markdown to Weblate is to move strings manually. If you are OK with switching to Weblate for Czech - let's do this exercise once.

**Reply by Grreby** [reply]
Ok, so should I copy the sentences from the Markdown file to appropriate strings in Weblate?

**Reply by biodranik** [reply]
CC @rtsisyk

**Reply by rtsisyk** [reply]
> Ok, so should I copy the sentences from the Markdown file to appropriate strings in Weblate?

@Grreby, correct. Luckily, this is the one-time operation. I've merged your partial Czech translation from Weblate into `master`, but generation of Markdowns is not enabled yet. Let's try to reach 80% or more, and then I will raise a PR with re-generated Markdowns.

**Comment by LaoshuBaby** [comment]
Shocking , because [I've forked a repo](https://github.com/LaoshuBaby/organicmaps.github.io/tree/translation_zh-Hans) since last month and tried to translate a few sentences. 
~~(I didn’t continue because I found out that the merge method of PR is Squash instead of Merge, I think this can obliterate edit history so I did not continue)~~ 
How to extract the diff part as the starting content and continue to work on weblate? Should I have a PR?

**Reply by rtsisyk** [reply]
@LaoshuBaby I will incorporate [your changes of `config.toml`](https://github.com/LaoshuBaby/organicmaps.github.io/commit/fe5e915fb8d4dedb48989a9583ba10143acd4d2a), no problem.

**Reply by rtsisyk** [reply]
Please continue on [Weblate](https://hosted.weblate.org/zen/organicmaps/website/zh_Hans/). 
Context: https://github.com/organicmaps/organicmaps.github.io/pull/85#issuecomment-1364163694

**Reply by rtsisyk** [reply]
I've merged zh_Hans partial translation into master. Markdowns will be generated after reaching 80%. Please keep going!

**Comment by nautilusx** [comment]
The process looks good. I've started the german translation. I'm sure I'll be done with it in the next few days. I hope we can continue with the strings of the app soon.
In the end weblate should be the only source for translations to avoid conflicts.

**Reply by biodranik** [reply]
Wait, the site is already translated into German: https://organicmaps.app/de/
Although the differences can be updated (e.g. a download header from the main page).

**Reply by rtsisyk** [reply]
> The process looks good. I've started the german translation. I'm sure I'll be done with it in the next few days. I hope we can continue with the strings of the app soon. In the end weblate should be the only source for translations to avoid conflicts.

Could you please compare this new Weblate experience vs old raw editing of Markdowns on GitHub. What is more convenient for you?

**Reply by nautilusx** [reply]
I know the the german translation of https://organicmaps.app/de/.
I use this, rather do a review and test the process.

Weblate is much better and easier for translations. In addition, there are machine quality checks and the glossary.

Weblate should become the only process and also the standard process for translations.

**Reply by biodranik** [reply]
How is it compared to Crowdin, where you can translate the whole Markdown file, without splitting it into chunks?

**Reply by rtsisyk** [reply]
@organicmaps/translations-de please weight in.

**Reply by rtsisyk** [reply]
Let's continue in https://github.com/organicmaps/organicmaps.github.io/pull/91

**Comment by nautilusx** [comment]
@biodranik and @rtsisyk Weblate is definitely the best platform for translations. I also use Crowdin and Transifex. They are not that good.

The German translation of the homepage is finished. I was able to correct a few small mistakes.

**Reply by rtsisyk** [reply]
https://github.com/organicmaps/organicmaps.github.io/pull/91

**Reply by nautilusx** [reply]
@rtsisyk Thanks for the great preview. I found 4 small mistakes. Can you please update the preview for double check?

**Comment by fitojb** [comment]
Thanks for providing l10n for the website on Weblate. I fixed typos in and proofread the Spanish translation and am about to start translating it to Catalan :)

**Reply by rtsisyk** [reply]
Re-based and merged. Thanks! Weblate is now switched to `master` branch and all changes should go seamlessly.

**Comment by nautilusx** [comment]
@rtsisyk and @biodranik please merged the latest strings from Weblate.

**Reply by rtsisyk** [reply]
Done. Let's continue in https://github.com/organicmaps/organicmaps.github.io/pull/91

**Comment by csidirop** [comment]
Hi, I'm a bit late and have already missed german , but would do greek over the next days. Unfortunately the How-To link doesn't lead anywhere:

> Useful links:
>  
>     * [Detailed instructions](https://github.com/organicmaps/organicmaps.github.io/blob/i18n/TRANSLATIONS.md#how-it-works)
> 

[TRANSLATIONS.md](https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md) is the old one. 
Also, the site seems to be under maintenance at the moment and/or really slow, so I can't look at it and try it out a bit. Anyways is there anything important to keep in mind?

Thanks :)

**Reply by rtsisyk** [reply]
> Hi, I'm a bit late and have already missed german , but would do greek over the next days. Unfortunately the How-To link doesn't lead anywhere:
> 
> > Useful links:
> > ```
> > * [Detailed instructions](https://github.com/organicmaps/organicmaps.github.io/blob/i18n/TRANSLATIONS.md#how-it-works)
> > ```
> 
> [TRANSLATIONS.md](https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md) is the old one. Also, the site seems to be under maintenance at the moment and/or really slow, so I can't look at it and try it out a bit. Anyways is there anything important to keep in mind?
> 
> Thanks :)

Fixed, thanks! i18n branch has been merged into master.

**Comment by ZeljkoBG** [comment]
@rtsisyk Does Organic translate to Weblate or not?
